core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -284,12 +289,18 @@ class KafkaController(val config : KafkaConfig, zkUtils: ZkUtils, val brokerStat
                   }
                   // If the broker is a follower, updates the isr in ZK and notifies the current leader
                   replicaStateMachine.handleStateChanges(Set(PartitionAndReplica(topicAndPartition.topic,
-                    topicAndPartition.partition, id)), OfflineReplica)
+                    topicAndPartition.partition, id)), OfflineReplica, newBatch = false)
                 }
               }
             }
           }
       }
+
+      inLock(controllerContext.controllerLock) {
```

Comment:
@lindong28, @jjkoshy : Yes, I agree that would fix the issue, but it feels a bit hacky. Fundamentally, the issue is that we have 2 independent threads (ZK event thread and Kafka request thread) that can change the partition state concurrently. Also, I think the bulk of the time spent in controlled shutdown is probably the synchronous writes (of the new leader) to ZK, which is not addressed in this patch. So, instead of doing incremental patches that potentially makes the code even more brittle, I am wondering if we should look into how to redesign the controller holistically (e.g., moving to single threaded model, async ZK access, etc).